ocaml/typing/jkind.ml

```diff
@@ -2116,37 +1792,383 @@ module Desc = struct
         }
 end
 
-let rec default_all_sort_variables_to_value (t : t) =
-  match t with
-  | Type ty -> Type.default_to_value ty
-  | Arrow { args; result } ->
-    List.iter default_all_sort_variables_to_value args;
-    default_all_sort_variables_to_value result
+let default_all_sort_variables_to_value (t : t) =
+  let rec go = function
+    | Type ty -> ignore (Type.Jkind_desc.default_to_value_and_get ty)
+    | Arrow { args; result } ->
+      List.iter go args;
+      go result
+  in
+  go t.jkind
 
-let get (t : t) : Desc.t =
-  match t with
-  | Type ty -> Type ty
-  | Arrow { args; result } -> Desc.Arrow { args; result }
+let get t : Desc.t =
+  match t.jkind with
+  | Type ty -> Type { jkind = ty; history = t.history; has_warned = false }
+  | Arrow { args; result } ->
+    Arrow
+      { args =
+          List.mapi
+            (fun i arg : t ->
+              { jkind = arg;
+                history =
+                  Projection
+                    { reason = Arrow_argument i;
+                      jkind = t.jkind;
+                      history = t.history
+                    };
+                has_warned = false
+              })
+            args;
+        result =
+          { jkind = result;
+            history =
+              Projection
+                { reason = Arrow_result; jkind = t.jkind; history = t.history };
+            has_warned = false
+          }
+      }
```

Comment:
Why are you adding the `projection` histories here rather than on construction? Most values are destructed more often than they are constructed.

Reply:
I think we had discussed this. It's because a `Jkind.t` is a full jkind with history, while its description does not have histories on its components. 
Sub-component `Jkind.t`s are only created for higher kinds, which are the (relatively) rare case, so this shouldn't be a performance problem. My hope was also that any unused parts will be optimised away.
`Jkind.get` is also usually used when at least one of the components will be used for another operation, and not just for any destruction.

I don't think I'm happy with a redesign that will make `Jkind.t` and `Jkind.Jkind_desc.t` mutually recursive on each other, as they would also have to be mutually recursive with `Jkind.Type.t` and `Jkind.Type.Jkind_desc.t` through `Jkind.History.t`. But I don't see if you actually mean another solution here?

Reply:
ad: above response
from our discussion, I'll be merging the definition of these types and things should work out fine.